File: protocol/bridge/common/src/main/java/ink/reactor/protocol/bridge/common/handler/SingleHandlerStorage.java
package ink.reactor.protocol.bridge.common.handler;

import ink.reactor.protocol.api.PlayerConnection;
import ink.reactor.protocol.api.buffer.reader.ReaderBuffer;
import ink.reactor.protocol.api.packet.PacketHandler;
import lombok.RequiredArgsConstructor;

import java.util.Arrays;

@RequiredArgsConstructor
public final class SingleHandlerStorage implements MutablePacketHandlerStorage {
    private final PacketHandler handler;

    @Override
    public void callHandlers(final PlayerConnection connection, final ReaderBuffer readerBuffer) {
        handler.handle(connection, readerBuffer);
    }

    @Override
    public MutablePacketHandlerStorage add(final PacketHandler handler) {
        return new ArrayHandlerStorage(new PacketHandler[] {this.handler, handler});
    }

    @Override
    public MutablePacketHandlerStorage remove(final PacketHandler handler) {
        return this.handler.equals(handler) ? null : this;
    }

    @Override
    public MutablePacketHandlerStorage add(final PacketHandler... handlers) {
        final PacketHandler[] packetHandlers = Arrays.copyOf(handlers, handlers.length + 1);
        packetHandlers[handlers.length] = handler;

        return new ArrayHandlerStorage(packetHandlers);
    }

    @Override
    public MutablePacketHandlerStorage remove(final PacketHandler... handlers) {
        return handlers.length == 0 ? this : remove(handlers[0]);
    }
}

File: networking/internal/src/main/kotlin/ink/reactor/network/internal/packet/storage/SingleHandlerStorage.kt
package ink.reactor.network.internal.packet.storage

import ink.reactor.network.api.packet.Packet
import ink.reactor.network.api.packet.handler.PacketHandler
import ink.reactor.network.api.player.PlayerConnection

internal class SingleHandlerStorage(private val handler: PacketHandler) : MutablePacketHandlerStorage {

    override fun callHandlers(connection: PlayerConnection, packet: Packet) {
        handler.handle(connection, packet)
    }

    override fun add(handler: PacketHandler): MutablePacketHandlerStorage {
        return ArrayHandlerStorage(arrayOf(this.handler, handler))
    }

    override fun remove(handler: PacketHandler): MutablePacketHandlerStorage? {
        return if (this.handler == handler) null else this
    }

    override fun add(vararg handlers: PacketHandler): MutablePacketHandlerStorage {
        val packetHandlers = handlers.toMutableList().apply { add(handler) }.toTypedArray()
        return ArrayHandlerStorage(packetHandlers)
    }

    override fun remove(vararg handlers: PacketHandler): MutablePacketHandlerStorage? {
        return if (handlers.isEmpty()) this else remove(handlers[0])
    }
}
